File: AnimalTheoryRoyale/Backend/Models/MapObstacles.cs
using System;
using System.Collections.Generic;

namespace Backend.Models
{
    public class Obstacle
    {
        public string Id { get; set; } = "";
        public string Type { get; set; } = ""; // "circle" or "box"
        public float X { get; set; }
        public float Z { get; set; }
        public float Radius { get; set; } // for circle
        public float Width { get; set; } // for box
        public float Depth { get; set; } // for box
        public float Rotation { get; set; } // for box
        public bool Blocking { get; set; }
        public bool IsWater { get; set; }
    }

    public static class MapObstacles
    {
        public const float MAP_SIZE = 1000f;
        public static readonly List<Obstacle> Obstacles = new List<Obstacle>();

        static MapObstacles()
        {
            GenerateObstacles();
        }

        private static void GenerateObstacles()
        {
            float Rng(ref long seed)
            {
                seed = (seed * 16807) % 2147483647;
                return (float)seed / 2147483647f;
            }

            long s = 42;
            float Rand() => Rng(ref s);

            // === WATER (Deep river) ===
            Obstacles.Add(new Obstacle { Id = "river_main", Type = "box", X = 0, Z = 0, Width = MAP_SIZE * 1.5f, Depth = 60, Rotation = MathF.PI / 4, Blocking = true, IsWater = true });

            // === BRIDGE ===
            Obstacles.Add(new Obstacle { Id = "bridge_rail_1", Type = "box", X = (float)(-7.5 * Math.Cos(Math.PI / 4)), Z = (float)(-7.5 * Math.Sin(Math.PI / 4)), Width = 1, Depth = 80, Rotation = MathF.PI / 4, Blocking = true });
            Obstacles.Add(new Obstacle { Id = "bridge_rail_2", Type = "box", X = (float)(7.5 * Math.Cos(Math.PI / 4)), Z = (float)(7.5 * Math.Sin(Math.PI / 4)), Width = 1, Depth = 80, Rotation = MathF.PI / 4, Blocking = true });

            // ROCKS
            for (int i = 0; i < 80; i++)
            {
                float ox = (Rand() - 0.5f) * MAP_SIZE * 0.85f;
                float oz = (Rand() - 0.5f) * MAP_SIZE * 0.85f;
                if (IsOnBridge(ox, oz)) continue;
                Obstacles.Add(new Obstacle { Id = $"rock_{i}", Type = "circle", X = ox, Z = oz, Radius = 1 + Rand() * 4, Blocking = true });
            }

            // TREES
            for (int i = 0; i < 60; i++)
            {
                float ox = (Rand() - 0.5f) * MAP_SIZE * 0.8f;
                float oz = (Rand() - 0.5f) * MAP_SIZE * 0.8f;
                if (IsOnBridge(ox, oz)) continue;
                float scale = 2 + Rand() * 3;
                Obstacles.Add(new Obstacle { Id = $"tree_{i}", Type = "circle", X = ox, Z = oz, Radius = 0.3f * scale, Blocking = true });
            }

            // BUILDINGS
            for (int i = 0; i < 15; i++)
            {
                float ox = (Rand() - 0.5f) * MAP_SIZE * 0.7f;
                float oz = (Rand() - 0.5f) * MAP_SIZE * 0.7f;
                if (IsOnBridge(ox, oz)) continue;
                float scale = 3 + Rand() * 5;
                Obstacles.Add(new Obstacle { Id = $"bldg_{i}", Type = "box", X = ox, Z = oz, Width = scale * 2, Depth = scale * 2, Rotation = Rand() * MathF.PI * 2, Blocking = true });
            }

            // HILLS
            for (int i = 0; i < 25; i++)
            {
                float ox = (Rand() - 0.5f) * MAP_SIZE * 0.9f;
                float oz = (Rand() - 0.5f) * MAP_SIZE * 0.9f;
                float radius = 20 + Rand() * 40;
                if (IsOnBridge(ox, oz)) continue;
                Obstacles.Add(new Obstacle { Id = $"hill_{i}", Type = "circle", X = ox, Z = oz, Radius = radius * 0.5f, Blocking = true });
            }
        }

        public static bool IsOnBridge(float px, float pz)
        {
            float cosA = MathF.Cos(-MathF.PI / 4);
            float sinA = MathF.Sin(-MathF.PI / 4);
            float localX = px * cosA - pz * sinA;
            float localZ = px * sinA + pz * cosA;

            return Math.Abs(localX) <= 8 && Math.Abs(localZ) <= 40;
        }

        public static bool IsPositionBlocked(float x, float z, float playerRadius = 2.5f)
        {
            if (Math.Abs(x) > MAP_SIZE / 2 - 2 || Math.Abs(z) > MAP_SIZE / 2 - 2) return true;

            foreach (var obs in Obstacles)
            {
                if (!obs.Blocking) continue;

                if (obs.Type == "circle")
                {
                    float dx = x - obs.X;
                    float dz = z - obs.Z;
                    float dist = MathF.Sqrt(dx * dx + dz * dz);
                    if (dist < obs.Radius + playerRadius) return true;
                }
                else if (obs.Type == "box")
                {
                    float rot = obs.Rotation;
                    float cosA = MathF.Cos(-rot);
                    float sinA = MathF.Sin(-rot);

                    float dx = x - obs.X;
                    float dz = z - obs.Z;

                    float localX = dx * cosA - dz * sinA;
                    float localZ = dx * sinA + dz * cosA;

                    float halfW = obs.Width / 2 + playerRadius;
                    float halfD = obs.Depth / 2 + playerRadius;

                    if (Math.Abs(localX) < halfW && Math.Abs(localZ) < halfD)
                    {
                        if (obs.IsWater && IsOnBridge(x, z)) continue;
                        return true;
                    }
                }
            }

            return false;
        }
    }
}

File: AnimalTheoryRoyale/Frontend/src/game3d/MapObstacles.js
// Shared map obstacles data for Frontend rendering and collision logic.

export const MAP_SIZE = 1000;

// Pseudo-random generator for consistent obstacle placement
const rng = (seed) => {
  let s = seed;
  return () => { s = (s * 16807) % 2147483647; return s / 2147483647; };
};
const rand = rng(42);

const obstacles = [];

// === WATER (Deep river) ===
// River is a box cutting diagonally
obstacles.push({
  id: 'river_main',
  type: 'box',
  x: 0,
  z: 0,
  width: MAP_SIZE * 1.5,
  depth: 60,
  rotation: Math.PI / 4, // 45 degrees
  blocking: true,
  isWater: true
});

// === BRIDGE ===
// Bridge floor allows walking over the water. 
// We model the bridge railings as blocking, and the floor as walkable (we just don't add the floor as an obstacle)
obstacles.push({
  id: 'bridge_rail_1',
  type: 'box',
  x: -7.5 * Math.cos(Math.PI / 4), // Rotate manually since simple collision won't handle rotated boxes easily without math
  z: -7.5 * Math.sin(Math.PI / 4),
  width: 1,
  depth: 80,
  rotation: Math.PI / 4,
  blocking: true
});
obstacles.push({
  id: 'bridge_rail_2',
  type: 'box',
  x: 7.5 * Math.cos(Math.PI / 4),
  z: 7.5 * Math.sin(Math.PI / 4),
  width: 1,
  depth: 80,
  rotation: Math.PI / 4,
  blocking: true
});

// Helper to check if a point is on the bridge
export function isOnBridge(px, pz) {
  // Bridge is at 0,0 rotated 45 degrees. Length 80, width 16.
  // Un-rotate the point
  const cosA = Math.cos(-Math.PI / 4);
  const sinA = Math.sin(-Math.PI / 4);
  const localX = px * cosA - pz * sinA;
  const localZ = px * sinA + pz * cosA;
  
  if (Math.abs(localX) <= 8 && Math.abs(localZ) <= 40) {
    return true;
  }
  return false;
}

// === ROCKS ===
for (let i = 0; i < 80; i++) {
  let ox = (rand() - 0.5) * MAP_SIZE * 0.85;
  let oz = (rand() - 0.5) * MAP_SIZE * 0.85;
  // Don't place on the bridge or river
  if (isOnBridge(ox, oz)) continue;

  obstacles.push({
    id: `rock_${i}`,
    type: 'circle',
    x: ox,
    z: oz,
    radius: 1 + rand() * 4,
    blocking: true,
    renderType: 'rock'
  });
}

// === TREES ===
for (let i = 0; i < 60; i++) {
  let ox = (rand() - 0.5) * MAP_SIZE * 0.8;
  let oz = (rand() - 0.5) * MAP_SIZE * 0.8;
  if (isOnBridge(ox, oz)) continue;

  let scale = 2 + rand() * 3;
  obstacles.push({
    id: `tree_${i}`,
    type: 'circle',
    x: ox,
    z: oz,
    radius: 0.3 * scale, // Trunk radius
    blocking: true,
    renderType: 'tree',
    renderScale: scale
  });
}

// === BUILDINGS ===
for (let i = 0; i < 15; i++) {
  let ox = (rand() - 0.5) * MAP_SIZE * 0.7;
  let oz = (rand() - 0.5) * MAP_SIZE * 0.7;
  if (isOnBridge(ox, oz)) continue;

  let scale = 3 + rand() * 5;
  obstacles.push({
    id: `bldg_${i}`,
    type: 'box',
    x: ox,
    z: oz,
    width: scale * 2,
    depth: scale * 2,
    rotation: rand() * Math.PI * 2,
    height: 4 + rand() * 8,
    blocking: true,
    renderType: 'building',
    renderScale: scale
  });
}

// === HILLS (Too steep to walk) ===
// We make the inner 50% of the hill blocking
for (let i = 0; i < 25; i++) {
  let ox = (rand() - 0.5) * MAP_SIZE * 0.9;
  let oz = (rand() - 0.5) * MAP_SIZE * 0.9;
  let radius = 20 + rand() * 40;
  let height = 4 + rand() * 12;
  
  if (isOnBridge(ox, oz)) continue;

  obstacles.push({
    id: `hill_${i}`,
    type: 'circle',
    x: ox,
    z: oz,
    radius: radius * 0.5, // Only the center/peak is blocking
    blocking: true,
    renderType: 'hill',
    renderRadius: radius,
    renderHeight: height
  });
}

export const MAP_OBSTACLES = obstacles;

// Basic collision logic for Frontend
export function isPositionBlocked(x, z, playerRadius = 2) {
  // Map bounds
  if (Math.abs(x) > MAP_SIZE / 2 - 2 || Math.abs(z) > MAP_SIZE / 2 - 2) return true;

  // Check each obstacle
  for (const obs of MAP_OBSTACLES) {
    if (!obs.blocking) continue;

    if (obs.type === 'circle') {
      const dx = x - obs.x;
      const dz = z - obs.z;
      const dist = Math.sqrt(dx * dx + dz * dz);
      if (dist < obs.radius + playerRadius) {
        return true;
      }
    } 
    else if (obs.type === 'box') {
      // Un-rotate player point around box center
      const rot = obs.rotation || 0;
      const cosA = Math.cos(-rot);
      const sinA = Math.sin(-rot);
      
      const dx = x - obs.x;
      const dz = z - obs.z;
      
      const localX = dx * cosA - dz * sinA;
      const localZ = dx * sinA + dz * cosA;
      
      const halfW = obs.width / 2 + playerRadius;
      const halfD = obs.depth / 2 + playerRadius;
      
      if (Math.abs(localX) < halfW && Math.abs(localZ) < halfD) {
        // Exception: Water block but we are on the bridge
        if (obs.isWater && isOnBridge(x, z)) {
          continue; // Allow moving over water IF on bridge
        }
        return true;
      }
    }
  }
  
  return false;
}
